tf/bonsai/bonsai.py

```diff
@@ -0,0 +1,210 @@
+import utils
+import tensorflow as tf
+import numpy as np
+
+## Bonsai Class
+class Bonsai:
+	## Constructor
+	def __init__(self, C, F, P, D, S, lW, lT, lV, lZ, 
+		sW, sT, sV, sZ, lr = None, W = None, T = None, 
+		V = None, Z = None, feats = None):
+
+		self.dataDimension = F + 1
+		self.projectionDimension = P
+		
+		if (C > 2):
+			self.numClasses = C
+		elif (C == 2):
+			self.numClasses = 1
+
+		self.treeDepth = D
+		self.sigma = S
+		
+		## Regularizer coefficients
+		self.lW = lW
+		self.lV = lV
+		self.lT = lT
+		self.lZ = lZ
+		
+		## Sparsity hyperparams
+		self.sW = sW
+		self.sV = sV
+		self.sT = sT
+		self.sZ = sZ
+
+		self.internalNodes = 2**self.treeDepth - 1
+		self.totalNodes = 2*self.internalNodes + 1
+
+		## The Parameters of Bonsai
+		self.W = self.initW(W)
+		self.V = self.initV(V)
+		self.T = self.initT(T)
+		self.Z = self.initZ(Z)
+
+		## Placeholders for Hard Thresholding and Sparse Training
+		self.Wth = tf.placeholder(tf.float32, name='Wth')
+		self.Vth = tf.placeholder(tf.float32, name='Vth')
+		self.Zth = tf.placeholder(tf.float32, name='Zth')
+		self.Tth = tf.placeholder(tf.float32, name='Tth')
+		
+		## Placeholders for Features and labels
+		## feats are to be fed when joint training is being done with Bonsai as end classifier
+		if feats is None:
+			self.x = tf.placeholder("float", [None, self.dataDimension])
+		else:
+			self.x = feats
+
+		self.y = tf.placeholder("float", [None, self.numClasses])
+
+		## Placeholder for batch size, needed for Multiclass hinge loss
+		self.batch_th = tf.placeholder(tf.int64, name='batch_th')
+
+		self.sigmaI = 1.0
+		if lr is not None:
+			self.learningRate = lr
+		else:
+			self.learningRate = 0.01
+
+		## Functions to setup required graphs
+		self.hardThrsd()
+		self.sparseTraining()
+		self.lossGraph()
+		self.trainGraph()
+		self.accuracyGraph()
+
+	## Functions to initilaise Params (Warm start possible with given numpy matrices)
+	def initZ(self, Z):
+		if Z is None:
+			Z = tf.random_normal([self.projectionDimension, self.dataDimension])
+		Z = tf.Variable(Z, name='Z', dtype=tf.float32)
+		return Z
+
+	def initW(self, W):
+		if W is None:
+			W = tf.random_normal([self.numClasses*self.totalNodes, self.projectionDimension])
```

Comment:
@adityakusupati I was wondering, should we allow a custom initializer as an optional argument instead of having a random normal initializer hardcoded? Like `LSTMCell` has?

Reply:
The input param W, can be thought of as a custom initialiser. You can generate any intended initialisation and load it with W.